cores/esp8266/core_esp8266_spi_utils.cpp

```diff
@@ -0,0 +1,144 @@
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2.1 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <stdint.h>
+#include <string.h>
+
+// register names
+#include "esp8266_peri.h"
+
+// for flashchip
+#include "spi_flash.h"
+
+// for PRECACHE_*
+#include "core_esp8266_features.h"
+
+#include "spi_utils.h"
+
+extern "C" uint32_t Wait_SPI_Idle(SpiFlashChip *fc);
+
+/*
+ * critical part of SPICommand.
+ * Kept in a separate function to aid with precaching
+ * PRECACHE_* saves having to make the function IRAM_ATTR.
+ *
+ * Note: if porting to ESP32 mosi/miso bits are set in 2 registers, not 1.
+ */
+static SpiOpResult PRECACHE_ATTR
+_SPICommand(volatile uint32_t spiIfNum,
+            uint32_t spic,uint32_t spiu,uint32_t spiu1,uint32_t spiu2,
+            uint32_t *data,uint32_t write_words,uint32_t read_words)
+{ 
+  if (spiIfNum>1) return SPI_RESULT_ERR;
+
+  // force SPI register access via base+offest. 
+  // Prevents loading individual address constants from flash.
+  uint32_t *spibase = (uint32_t*)(spiIfNum ? &(SPI1CMD) : &(SPI0CMD));
+  #define SPIREG(reg) *((volatile uint32_t *)(spibase+(&(reg) - &(SPI0CMD))))
+
+  // preload any constants and functions we need into variables
+  // Everything must be volatile or the optimizer can treat them as 
+  // constants, resulting in the flash reads we're trying to avoid
+  void    *(* volatile memcpyp)(void *,const void *, size_t) = memcpy;
+  uint32_t (* volatile Wait_SPI_Idlep)(SpiFlashChip *) = Wait_SPI_Idle;
+  volatile SpiFlashChip *fchip=flashchip;
+  volatile uint32_t spicmdusr=SPICMDUSR;
+
+  if (!spiIfNum) {
+     // Only need to precache when using SPI0
+     PRECACHE_START();
+     Wait_SPI_Idlep((SpiFlashChip *)fchip);
+  }
+  
+  uint32_t old_spi_usr = SPIREG(SPI0U);
+  uint32_t old_spi_usr2= SPIREG(SPI0U2);
+  uint32_t old_spi_c   = SPIREG(SPI0C);
+
+  //SPI0S &= ~(SPISE|SPISBE|SPISSE|SPISCD);
+  SPIREG(SPI0C) = spic;
+  SPIREG(SPI0U) = spiu;
+  SPIREG(SPI0U1)= spiu1;
+  SPIREG(SPI0U2)= spiu2;
+
+  if (write_words>0) {
+     // copy the outgoing data to the SPI hardware
+     memcpyp((void*)&(SPIREG(SPI0W0)),data,write_words*4);
+  }
+
+  // Start the transfer
+  SPIREG(SPI0CMD) = spicmdusr;
+
+  // wait for the command to complete
+  uint32_t timeout = 1000;
+  while ((SPIREG(SPI0CMD) & spicmdusr) && timeout--) {}
```

Comment:
Pleade either replace {} with ; or put the {} on its own line

Reply:
The timeout here is basically "retry up to 1000 times". Can there be longer-running transfers where the 1000 loop limit is reached? What about flash speed (i. e. lower bus freq) or inherently slower flash chips?

Reply:
The code I based this on didn't implement a timeout at all - it just polled the state until it changed, or forever if it didn't.
Testing showed that typical completion times only decremented timeout by 2-3 at most with 40mhz flash. Even at 1mhz SPI bus speeds, the timeout is never going to be even close to being reached, but I figured it was better than relying on a WDT reset if something went wrong.